zebra-chain/src/value_balance.rs

```diff
@@ -0,0 +1,248 @@
+//! A type that can hold the Zcash four types of value pools
+
+use crate::amount::{Amount, Constraint, Error, NegativeAllowed, NonNegative};
+
+/// A structure that hold amounts for the four value pool types.
+#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(bound = "C: Constraint")]
+pub struct ValueBalance<C = NegativeAllowed> {
+    transparent: Amount<C>,
+    sprout: Amount<C>,
+    sapling: Amount<C>,
+    orchard: Amount<C>,
+}
+
+impl<C> ValueBalance<C>
+where
+    C: Constraint + Copy,
+{
+    /// [Consensus rule]: The remaining value in the transparent transaction value pool MUST
+    /// be nonnegative.
+    ///
+    /// This rule applies to Block and Mempool transactions.
+    ///
+    /// [Consensus rule]: https://zips.z.cash/protocol/protocol.pdf#transactions
+    pub fn remaining_transaction_value(&self) -> Result<Amount<NonNegative>, Error> {
+        // This rule checks the transparent value balance minus the sum of the sprout,
+        // sapling, and orchard value balances in a transaction is nonnegative.
+        (self.transparent - (self.sprout + self.sapling + self.orchard)?)?
+            .constrain::<NonNegative>()
+    }
+
+    /// Creates a `ValueBalance` from the given transparent amount.
+    pub fn from_transparent_amount(transparent_amount: Amount<C>) -> Self {
+        ValueBalance {
+            transparent: transparent_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Creates a `ValueBalance` from the given sprout amount.
+    pub fn from_sprout_amount(sprout_amount: Amount<C>) -> Self {
+        ValueBalance {
+            sprout: sprout_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Creates a `ValueBalance` from the given sapling amount.
+    pub fn from_sapling_amount(sapling_amount: Amount<C>) -> Self {
+        ValueBalance {
+            sapling: sapling_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Creates a `ValueBalance` from the given orchard amount.
```

Comment:
Nit: I'd add intra-doc links as a minor helper shortcut.

```suggestion
    /// Creates a [`ValueBalance`] from the given transparent amount.
    pub fn from_transparent_amount(transparent_amount: Amount<C>) -> Self {
        ValueBalance {
            transparent: transparent_amount,
            ..Default::default()
        }
    }

    /// Creates a [`ValueBalance`] from the given sprout amount.
    pub fn from_sprout_amount(sprout_amount: Amount<C>) -> Self {
        ValueBalance {
            sprout: sprout_amount,
            ..Default::default()
        }
    }

    /// Creates a [`ValueBalance`] from the given sapling amount.
    pub fn from_sapling_amount(sapling_amount: Amount<C>) -> Self {
        ValueBalance {
            sapling: sapling_amount,
            ..Default::default()
        }
    }

    /// Creates a [`ValueBalance`] from the given orchard amount.
```